Extract divide into a new IntegralDomain class

hseg opened the pull request:
Closes #62. Based on #64, though if that is rejected can be CPPifyed

Unfortunately, to still be able to provide default implementations we have
import cycles. These can be avoided by moving `IntegralDomain` into
Euclidean.hs, though it is unclear that the name would then still be
appropriate.

Cycles caused mainly by:
* Use of `WrappedIntegral`/`WrappedFractional` for `DerivingVia`
  instances for `IntegralDomain`
* `GcdDomain` now needs `IntegralDomain` context

In addition, to minimize context, added `Fractional (Complex a)` constraint to
the `Complex a` instances, which GHC dislikes. If the required extensions are
unpalatable, the constraint needs to be replaced with a `RealFrac a` constraint.

Not all types got a `IntegralDomain` instance, since I couldn't work out
what they should be. These include the 'containers' types and `IO`.

hseg committed:
Extract divide into a new IntegralDomain class

Unfortunately, to still be able to provide default implementations we
have import cycles. These can be avoided by moving `IntegralDomain` into
Euclidean.hs, though it is unclear that the name would then still be
appropriate.

Cycles caused mainly by:
* Use of `WrappedIntegral`/`WrappedFractional` for `DerivingVia`
  instances for `IntegralDomain`
* `GcdDomain` now needs `IntegralDomain` context

In addition, to minimize context, added `Fractional (Complex a)`
constraint to the `Complex a` instances, which GHC dislikes.
If the required extensions are unpalatable, the constraint needs to be
replaced with a `RealFrac a` constraint.

Not all types got a `IntegralDomain` instance, since I couldn't work out
what it should be. These include the 'containers' types and `IO`.

hseg commented:
Testsuite fails due to `quickcheck-classes` expecting old `Euclidean` interface.
Adding PR there to update expectations.

Bodigrim commented:
Let's really keep it orthogonal to #64, otherwise it becomes difficult to review actual changes.

hseg commented:
OK, will rebase onto master and reimplement instance derivation in CPP.
Currently fixing the testsuite + instances to make build pass, will rebase once that's green.
Though would be happier with a CPP-free version.

Bodigrim commented:
@hseg could you please rebase? Just to make relevant changes more clearly visible.

chessai commented:
ping @hseg would you like to continue working on this?

hseg commented:
Sorry, had a long enough thesis crunch that the patch slipped my mind. Can try seeing if I can bring it back into shape this week, don't like leaving loose ends.

On 28 August 2021 1:22:55 am GMT+03:00, chessai ***@***.***> wrote:
>ping @hseg would you like to continue working on this?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/chessai/semirings/pull/65#issuecomment-907506211

chessai commented:
> Sorry, had a long enough thesis crunch that the patch slipped my mind. Can try seeing if I can bring it back into shape this week, don't like leaving loose ends.
> 
> On 28 August 2021 1:22:55 am GMT+03:00, chessai ***@***.***> wrote:
> >ping @hseg would you like to continue working on this?
> >
> >-- 
> >You are receiving this because you were mentioned.
> >Reply to this email directly or view it on GitHub:
> >https://github.com/chessai/semirings/pull/65#issuecomment-907506211

Understandable. If you need more time no problem, I can always just ping again at a later date.

hseg commented:
Rebased the original patch, but it isn't ready yet.
Going by my remarks in #62, it appears I'll need to patch https://github.com/andrewthad/quickcheck-classes for the testsuite to work as well.
Am going through the old code to try and recall what I did.
Some points:
* Are we OK with `FlexibleContexts`, `UndecideableInstances`? Should be benign
* Note that there are import cycles because `GcdDomain` now needs an
  `IntegralDomain` context, and because `Wrapped{Integral,Fractional}` are
  defined in `Euclidean.hs`
* Rewrote the `Complex a` instances to require `Fractional (Complex a)` rather
  than the equivalent `RealFloat a`, since I find it more readable. That's
  probably what's pulling in the `FlexibleContexts`, and tastes may vary.
  Note as well that this infects other `Complex a` instances with this
  `Fractional (Complex a)` context
* Some types did not get `IntegralDomain` instances since I couldn't figure out
  a legal instance for them, in particular the `containers` types and `IO`.
* Undocumented added instance: `Ring Proxy`
* TH code is a little ugly -- in particular am concerned by the lack of
  type-safety around the types passed to the splices and the lack of hygiene in
  the generation of a fresh type variable

The import cycles in particular suggest we may want to have a clearer picture of
what goes where.

If the above isn't a concern, am pushing a couple final edits to my branch, then
will work on `quickcheck-classes` and updating the testsuite over the weekend.

Bodigrim commented:
I'm afraid I'm still (see discussion in #62) unconvinced that this change bears its weight. Changing the hierarchy of type classes breaks all existing users of `semirings` package, which has been otherwise stable for several years. Not the end of the world, but not something to do on a whim as well. Could we possibly have a proposal, clearly describing pros/cons and migration path?

hseg commented:
Right, fair enough, that was my understanding at the end of #62 as well.
Note that my original usecase for opening #62 was resolved without the patch regardless.
Would shelve this at least until the modules ever get reorganised, if ever. Considering this library is the low-dependency/low-complexity version of `algebra`, I'm not convinced this Procrustean exercise suits it.

chessai commented:
Closing for now, feel free to re-open if there's any further work
